Allocation: introduce a new decider that balances the index shard count among nodes

zhubotang-wq opened the pull request:
For a cluster with n data nodes to host an index with m shards, each node ideally should host not significantly more than m / n shards each. This new allocation decider acts on this principle to respond with a NOT_PREFERRED in event of a node being allocated more shards than threshold.

Nodes in shutdown are excluded when computing fair workload for nodes. A load skew tolerance setting is added to permit nodes to own more than ideal number of shards for the index. 

Relates ES-12080

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Merge remote-tracking branch 'upstream/main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

zhubotang-wq committed:
Update docs/changelog/135875.yaml

Commit:
[CI] Auto commit changes from spotless

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

DiannaHohensee reviewed (commented):
I took a first pass. I didn't have time to take a peek at the testing yet.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

ywangd commented:
It would be helpful if you coud share the buildscan link for the failed test so that we have a clear context. Without it, I assume the failure is from [FullClusterRestartIT.testResize](https://github.com/elastic/elasticsearch/blob/80b4d97884465f3bf7f480786487cda8703e4247/qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/FullClusterRestartIT.java#L1546)

> this new shard count is not available in Cluster State

This is _not_ true. The original index is created with [3 primaries and 1 replica](https://github.com/elastic/elasticsearch/blob/80b4d97884465f3bf7f480786487cda8703e4247/qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/FullClusterRestartIT.java#L1549C63-L1549C64). The error message you shared above says

> For index [[testresize_split/UqXRWqRfRNSrBFZ_Yhzqfw]] with [6] shards,

Note the `6 shards` actually means 6 _primary_ shards, that is double the original number of `3` due to splitting. This is because the PR [computes](https://github.com/elastic/elasticsearch/pull/135875/files#diff-cafa177dbd45fe378559eee818f2b4f0bcd6dfbd372dff83b3b8badbdf6007abR72) the number as

```
var totalShards = allocation.getClusterState().routingTable(ProjectId.DEFAULT).index(index).size();
```

This is counting the size of a `IndexRoutingTable` which is determined by its array of `IndexShardRoutingTable`. The `IndexShardRoutingTable` in turn manages all copies (primary and replicas) of the same shard. Therefore the total number of shards (primaries and replicas) should be `IndexRoutingTable.size() * IndexShardRoutingTable.size() = 12`. But you might as well use `IndexMetadata.getTotalNumberOfShards()` for it.

I think using total number of shards makes sense for stateful. But in stateless, we probably should differentiate between primaries and replicas since their corresponding nodes are two disjoint sets (not sure what's the plan for this PR, I can see this as a 2nd step). 

Also, IIUC,the new decider does not return `NO`. So why would shards remain unassigned when it says `NOT_PREFERRED`. That does not sound right, especially for newly created shards.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexShardCountAllocationDecider.java:
```diff
@@ -0,0 +1,124 @@
+
+    public static final String NAME = "index_shard_count";
+
+    private final IndexShardCountConstraintSettings indexShardCountConstraintSettings;
```
Ack. fixed.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

In a balanced allocation, for an index with n shards on a cluster
of m nodes, each node should host not significantly more than n / m
shards. This decider enforces this principle.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

fix test regressions.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

Commit:
[CI] Auto commit changes from spotless

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,132 @@
+            return Decision.single(Decision.Type.YES, NAME, "Decider is disabled.");
+        }
+
+        // Never reject allocation of an unassigned shard
```
Is this still a WIP? Can this be undone now that Nick [fixed](https://github.com/elastic/elasticsearch/commit/30d45732b6e6508ef28e3e8e0c953e89c1ebd0af) the bug you encountered?

We want to spread out shards of newly created indices, to avoid hot-spots, so this doesn't seem like a good idea.

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,132 @@
+    public IndexBalanceAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.indexBalanceConstraintSettings = new IndexBalanceConstraintSettings(clusterSettings);
+        setClusterExcludeFilters(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING.getAsMap(settings));
+        clusterSettings.addAffixMapUpdateConsumer(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING, this::setClusterExcludeFilters, (a, b) -> {});
```
Agreed.

For a hot second I was thinking maybe ignore the settings and answer NOT_PREFERRED, but that would dilute the meaning of any other decider's NOT_PREFERRED responses.

> So taking care of shutdown nodes and index/search shard/node separation (as a follow-up) should be good for sometime to come.

The change isn't much use to us without shutdown and spreading out the primary shards. If we need to make this a serverless only decider, that's an option.

zhubotang-wq replied:
Ack. Removing this from available node filter.

ywangd replied:
Not sure if I follow, if the decider simply return `YES` when there is any exclusion settings, is that not sufficient for existing tests?

ywangd replied:
Yes it will subscribe the settings but only need to keep a `true` and `false` for each group without applying them. I think that is OK. My original point is to avoid the logic to computing a subset of candidate nodes which is more complexity and not useful in stateless. 

My concern is that applying a subset of the allocation filters is a hard to explain feature. It is much easier to say that this decider takes no action when there is any allocation filtering instead of having to explain that some allocation filters are considered but not others and why. I also think it is more like a half solution since it can return `NOT_PREFERRED` when there are filters other than the node level excludes.

zhubotang-wq replied:
Fixed.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

fix test regressions.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Convert decider to stateless.

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Convert decider to stateless.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Add unit tests

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Add unit tests

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Add unit tests

Commit:
[CI] Auto commit changes from spotless

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Add unit tests

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Add unit tests

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Add unit tests

zhubotang-wq committed:
Allocation: Include index shard counts as a criteria

Add unit tests

Commit:
[CI] Auto commit changes from spotless

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
fix test regressions

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+        }
+
+        if (isStateless == false) {
+            return Decision.single(Decision.Type.YES, NAME, "Decider does not currently support stateful.");
```
```suggestion
            return Decision.single(Decision.Type.YES, NAME, "Decider is disabled.");
```
Users of regular elasticsearch are going to find 'stateful' confusing. Let's keep it simple.

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+        }
+
+        assert node.node() != null;
+        assert node.node().getRoles() != null && node.node().getRoles().isEmpty() == false;
```
I don't think we need the `getRoles()` asserts.

`getRoles()` is not nullable, and if `getRoles()` is empty, the node doesn't have the roles we want and we return `YES`.

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+        assert node.node() != null;
+        assert node.node().getRoles() != null && node.node().getRoles().isEmpty() == false;
+        if (node.node().getRoles().contains(INDEX_ROLE) == false && node.node().getRoles().contains(SEARCH_ROLE) == false) {
+            return Decision.single(Decision.Type.YES, NAME, "Node has neither index nor search roles, outside purview.");
```
```suggestion
            return Decision.single(Decision.Type.YES, NAME, "Node has neither index nor search roles.");
```

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+                currentAllocation,
+                nomenclature,
+                index,
+                nomenclature
```
There is a lot of repeated information in this string. How about tightening it up to something like
```
"""
    There are [5] eligible nodes in the [index] tier for assignment of [4] shards in index [idx]. Ideally no more than [2]
    shard would be assigned per node (the index balance skew setting is [1.5]). This node is already assigned [3] shards of
    the index.
"""
```
Deduplicating the parameter list to
```
    eligibleNodes.size(),
    nomenclature,
    totalShards,
    index,
    idealAllocation,
    indexBalanceConstraintSettings.getLoadSkewTolerance(),
    currentAllocation
```

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+            null,
+            System.nanoTime()
+        );
+        routingAllocation.setDebugMode(RoutingAllocation.DebugMode.ON);
```
Why?

zhubotang-wq replied:
If I recall correctly, when this setting is not enabled, the explanations are not available in the response Decision.

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+    public void testCanAllocate_under_threshold() {
+        setup(true);
+
+        Settings settings = Settings.builder()
```
L195-L215 is duplicate code across tests. Since you have the `setup` method, can you deduplicate the test code?

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+
+        if (isStateless == false) {
+            return Decision.single(Decision.Type.YES, NAME, "Decider does not currently support stateful.");
+        }
```
Yes, the priority is serverless. In stateful we will need to access the data tier logic in the x-pack extension, to make things work correctly.

DiannaHohensee reviewed (commented):
Posting some feedback

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+            ShardId lastPrimaryShardId = new ShardId(indexMetadata.getIndex(), numberOfPrimaryShards - 1);
+            masterPrimaryShardRouting = TestShardRouting.newShardRouting(
+                lastPrimaryShardId,
+                masterNode.getId(),
```
This test setup was primarily designed to cover a case where the routing node has neither index nor search role but hosts this index, for which case the decider should recuse itself. Admittedly, using the master node for this purpose is not very realistic in stateless scenario, as suggested, this state may not exist in practice.

zhubotang-wq committed:
Update server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

Co-authored-by: Dianna Hohensee <artemisapple@gmail.com>

zhubotang-wq committed:
Update server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

Co-authored-by: Dianna Hohensee <artemisapple@gmail.com>

zhubotang-wq committed:
Update server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

Co-authored-by: Dianna Hohensee <artemisapple@gmail.com>

henningandersen reviewed (commented):
Only looked at the decider, looks good. I wonder if we also want can-remain, but assume this is a separate task. For instance with 3 shards on 2 nodes - then we do a scale up and want to get the one shard moved off one of the nodes.

zhubotang-wq commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderIT.java:
```diff
@@ -166,7 +166,11 @@ public void testHighNodeWriteLoadPreventsNewShardAllocation() {
+                Settings.builder()
+                    .put("cluster.routing.allocation.exclude._name", harness.firstDataNodeName)
+                    .put("cluster.routing.allocation.index_balance_decider.load_skew_tolerance", 1000)
+            );
```
Fixed.

zhubotang-wq commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+
+    private void collectEligibleNodes(RoutingAllocation allocation, Set<DiscoveryNode> eligibleNodes, DiscoveryNodeRole role) {
+        for (DiscoveryNode discoveryNode : allocation.nodes()) {
+            if (discoveryNode.canContainData()
```
Fixed.

zhubotang-wq committed:
fix test regressions

zhubotang-wq committed:
Merge branch '12080-index-shard-count-allocation-decider' of github.com:zhubotang-wq/elasticsearch into 12080-index-shard-count-allocation-decider

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+        }
+    }
+
+    public void testCanAllocate_under_threshold() {
```
Fixed.

zhubotang-wq committed:
fix test regressions

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+
+    private void setup(boolean allowMaster) {
+        numberOfPrimaryShards = allowMaster ? 11 : 10;
+        replicationFactor = 2;
```
Number of primary shards has been changed to random. Replication factors is limited by 
`assert distinctNodes(shards) : "more than one shard with same id assigned to same node (shards: " + shards + ")";`

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+
+        var routingAllocation = new RoutingAllocation(
+            null,
+            RoutingNodes.immutable(clusterState.globalRoutingTable(), clusterState.nodes()),
```
Fixed.

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+    private final Map<DiscoveryNode, List<ShardRouting>> nodeToShardRoutings = new HashMap<>();
+    private ClusterState clusterState;
+    private IndexMetadata indexMetadata;
+    private ShardRouting masterPrimaryShardRouting;
```
Fixed.

zhubotang-wq committed:
fix test regressions

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,358 @@
+
+        for (RoutingNode routingNode : List.of(routingSearchNodeOne, routingSearchNodeTwo)) {
+            assertDecisionMatches(
+                "Assigning a primary shard to a search node should succeed",
```
A typo. removed.

zhubotang-wq committed:
fix test regressions

zhubotang-wq committed:
fix test regressions

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
fix test regressions

zhubotang-wq committed:
Merge branch '12080-index-shard-count-allocation-decider' of github.com:zhubotang-wq/elasticsearch into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,147 @@
+        assert totalShards > 0;
+
+        final double idealAllocation = Math.ceil((double) totalShards / eligibleNodes.size());
+        final int threshold = (int) Math.ceil(idealAllocation) + indexBalanceConstraintSettings.getLoadSkewTolerance();
```
I am not sure what the extra `Math.ceil` brings us here?

zhubotang-wq replied:
The extra math.ceil() removed. Indeed it was unnecessary.

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,326 @@
+        settings = Settings.builder()
+            .put("stateless.enabled", "true")
+            .put("cluster.routing.allocation.index_balance_decider.enabled", "true")
+            .put("cluster.routing.allocation.index_balance_decider.load_skew_tolerance", 0)
```
I think this is default now, though ok to keep this line.

But should we have some verification that the skew tolerance works as intended?

henningandersen reviewed (commented):
Still looks good, left a few suggestions.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
fix test regressions

DiannaHohensee commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderIT.java:
```diff
@@ -0,0 +1,189 @@
+    private TestHarness setUpThreeHealthyDataNodesAndVerifyIndexShardsBalancedDistributed() {
+        Settings settings = Settings.builder()
+            .put(IndexBalanceConstraintSettings.INDEX_BALANCE_DECIDER_ENABLED_SETTING.getKey(), true)
+            .put(IndexBalanceConstraintSettings.INDEX_BALANCE_DECIDER_LOAD_SKEW_TOLERANCE.getKey(), 0)
```
No need to override default setting?

zhubotang-wq replied:
fixed.

DiannaHohensee commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderIT.java:
```diff
@@ -0,0 +1,189 @@
+            .build();
+        internalCluster().startMasterOnlyNode(settings);
+
+        final var dataNodes = internalCluster().startDataOnlyNodes(3, settings);
```
Perhaps Yang alluded to this, but the new decider isn't going to take action on a regular stateful data node, right? Was the suggestion to move IT testing into serverless?

zhubotang-wq replied:
Continuing this thread in [4755](https://github.com/elastic/elasticsearch-serverless/pull/4755).

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,326 @@
+    private RoutingAllocation routingAllocation;
+    private IndexBalanceAllocationDecider indexBalanceAllocationDecider;
+
+    private void setup(boolean exceedThreshold) {
```
Continuing my other comment thread, I'm hoping you can remove this boolean flag. I don't think you need to assign shards to ML / master nodes to test canAllocate.

DiannaHohensee reviewed (commented):
Took another look through focusing on the testing aspects.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Update server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java

Co-authored-by: Henning Andersen <33268011+henningandersen@users.noreply.github.com>

Commit:
[CI] Auto commit changes from spotless

zhubotang-wq commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        if (node.node().getRoles().contains(SEARCH_ROLE) && shardRouting.primary()) {
+            return Decision.single(Decision.Type.YES, NAME, "Decider allows primaries move to search nodes.");
```
Fixed.

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,326 @@
+        );
+
+        int ideal = numberOfPrimaryShards / 2;
+        int current = numberOfPrimaryShards / 2;
```
I am inclined to keep them as they are since they represent two distinct factors for this decider.

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,326 @@
+            indexShardRoutingBuilderMLNode.addShard(lastShardRouting);
+            indexRoutingTableBuilder.addIndexShard(indexShardRoutingBuilderMLNode);
+            routingTableBuilder.add(indexRoutingTableBuilder.build());
+            routingMachineLearningNode = RoutingNodesHelper.routingNode(machineLearningNode.getId(), machineLearningNode, lastShardRouting);
```
This setup is done so that for this routing node  `.hasIndex()` returns true but `getRoles().contain(SEARCH_ROLE)` and `getRoles().contains(INDEX_ROLE)` returns false to verify decider does not adjudicate on this case.

zhubotang-wq committed:
Address feedbacks

Commit:
[CI] Auto commit changes from spotless

zhubotang-wq commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,147 @@
+        if (currentAllocation >= threshold) {
+            String explanation = Strings.format(
+                "There are [%d] eligible nodes in the [%s] tier for assignment of [%d] shards in index [%s]. Ideally no more than [%.0f] "
+                    + "shard would be assigned per node (the index balance skew setting is [%d]). This node is already assigned [%d] shards"
```
Terminology updated.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Address feedbacks

zhubotang-wq committed:
Merge branch '12080-index-shard-count-allocation-decider' of github.com:zhubotang-wq/elasticsearch into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Update docs/changelog/135875.yaml

zhubotang-wq committed:
Address feedbacks

zhubotang-wq committed:
Merge remote-tracking branch 'origin/12080-index-shard-count-allocation-decider' into 12080-index-shard-count-allocation-decider

Commit:
[CI] Auto commit changes from spotless

zhubotang-wq committed:
Address feedbacks

zhubotang-wq committed:
Merge remote-tracking branch 'origin/12080-index-shard-count-allocation-decider' into 12080-index-shard-count-allocation-decider

zhubotang-wq commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderIT.java:
```diff
@@ -0,0 +1,189 @@
+        logger.info("---> Remove shard assignments of node " + testHarness.firstDataNodeName + " by excluding first data node.");
+        updateClusterSettings(Settings.builder().put("cluster.routing.allocation.exclude._name", testHarness.firstDataNodeName));
+
+        refreshClusterInfo();
```
Moving IndexBalanceAllocationDeciderIT to serverless repo. Migrating this PR comment to [4755](https://github.com/elastic/elasticsearch-serverless/pull/4755)

zhubotang-wq commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderIT.java:
```diff
@@ -0,0 +1,189 @@
+            if (indexRoutingTable == null) {
+                return false;
+            }
+            return checkShardAssignment(
```
Continuing this thread in [4755](https://github.com/elastic/elasticsearch-serverless/pull/4755).

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,326 @@
+                "An index node cannot own search shards. Decider inactive."
+            );
+        }
+
```
The incorrect assertDecisionMatches() description messages are fixed.

zhubotang-wq committed:
Address feedbacks

zhubotang-wq committed:
Address feedbacks

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Address feedbacks

zhubotang-wq committed:
Merge remote-tracking branch 'origin/12080-index-shard-count-allocation-decider' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Address feedbacks

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Address feedbacks

Commit:
[CI] Auto commit changes from spotless

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexBalanceConstraintSettings.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * This setting permits nodes to host more than ideally balanced number of index shards.
+     * Maximum tolerated index shard count = ideal + skew_tolerance
+     * i.e. ideal = 4 shards, skew_tolerance = 1
```
replace 'skew' in comment

zhubotang-wq replied:
I am inclined to keep this unchanged see skew tolerance is synonymous to excess shards.

zhubotang-wq commented:
@repantis . 

I’ve noticed a  pattern during this code review that’s impacting my ability to deliver on time. Often, the review process begins with a large number of surface-level comments — things like variable naming, spacing, line wrapping, or stylistic preferences — and only after a significant delay do I receive feedback on more critical aspects like design choices or testing strategy.

In this particular PR, there are a number of significant feedbacks that are particular relevant to this Code change 

- decider handling of tolerance of excess shards
- difference between stateful and stateless 
- Integration testing of this decider is affected by balancedShardAllocator built-in behaviour of considering shard weight. 

However, it is really hard to focus on these critical points with them arriving late. 

While I absolutely appreciate attention to detail and code quality, the current sequence and volume of non-blocking feedback tends to slow down iterations and split my focus unnecessarily. In some cases, it delays meaningful improvements that would actually benefit the project.

I’d love to discuss whether we can calibrate our review priorities a bit — for example, by encouraging reviewers to flag major design/test issues earlier in the review cycle, and possibly batch minor nitpicks separately or post-approval. This way, we can preserve code quality and maintain velocity.

I’m concerned that this pattern might also create friction or inefficiency across the team — especially when trying to keep up with release deadlines or switch contexts frequently due to nit-level distractions.

As the team grows, the new comers are bound to have many unknown unknown, thus it is vital to keep the criticial feedbacks earlier and unhindered. This minimize iterations and mitigate the risk of late discovery of critical flaws in implementations. 

CC @henningandersen .

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,352 @@
+    private DiscoveryNode indexNodeTwo;
+    private DiscoveryNode searchNodeOne;
+    private DiscoveryNode searchNodeTwo;
+    private DiscoveryNode masterNode;
```
unused

zhubotang-wq replied:
I will keep this unchanged due the presence of master node makes the test realistic.

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,352 @@
+    private ShardRouting indexTierShardRouting;
+    private ShardRouting searchTierShardRouting;
+    private List<RoutingNode> indexTier;
+    private List<RoutingNode> searchIier;
```
Typo: searchTier. T instead of I

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,352 @@
+            );
+        }
+
+        verifyCanAllocate();
```
This seems like the excess shards testing. Unlike the prior checks. Confusing to have the prior checks under testCanAllocateUnderThresholdWithExcessShards.

Maybe make two tests? testCanAllocateUnderThresholdWithExcessShards would have the verifyCanAllocate logic. And a second test, like testAllocateNewIndex.

DiannaHohensee commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+        }
+
+        assert node.node() != null;
+        assert node.node().getRoles().contains(INDEX_ROLE) || node.node().getRoles().contains(SEARCH_ROLE);
```
```suggestion
        assert node.node().getRoles().contains(INDEX_ROLE) || node.node().getRoles().contains(SEARCH_ROLE)
            : "Unexpected role found: " + node.node().getRoles();
```

DiannaHohensee reviewed (commented):
This is in good shape, pretty much done. I should be able to approve it once the new change requests are in.

I think you'll need to remove the `stateful-linked` label from https://github.com/elastic/elasticsearch-serverless/pull/4755, otherwise commit of this PR will automatically commit the linked serverless PR, and it's not ready yet. See this auto-comment on the [other PR](https://github.com/elastic/elasticsearch-serverless/pull/4755#issuecomment-3458661274).

Could you file a follow up JIRA ticket to add shutting down nodes and odd number of index shards to exercise the Math.ceil logic? I don't want to delay this code, but we should add that coverage next.

P.S. Could you leave resolving the comments I open to me? I'd like to check the resolution of a comment thread, and it's hard to open them again. And ongoing conversation threads should not be resolved until the discussion is finished.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+            nomenclature = "search";
+        }
+
+        assert eligibleNodes.isEmpty() == false;
```
If I’m remembering correctly, this assertion was introduced based on earlier review guidance. The new feedback seems to take a different position, and I’m concerned the shifting expectations may be introducing avoidable delays. Could we clarify the preferred direction?

zhubotang-wq commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    public void testCanAllocateHasDiscoveryNodeFilters() {
+        setup(true, randomBoolean());
```
Previously the deliberate test setup was flagged for lacking randomness and reducing robustness. After introducing randomness, the new feedback suggests reverting to the deliberate setup. I may be misunderstanding the intent, so could we clarify the testing standard? The guidance feels inconsistent to me.

DiannaHohensee replied:
> Previously the deliberate test setup was flagged for lacking randomness and reducing robustness. After introducing randomness, the new feedback suggests reverting to the deliberate setup.

Randomness is nice when it is expected not to affect the outcome of the test. It helps prove we don't have unexpected dependencies.

In this case, a random leniency in the number of shards allowed on the node will make your test randomly ineffectual / prove nothing. The test verifies that in the presence of filters, the Decider will always return YES. So the test must ensure a scenario where the Decider normally returns NO, but the filters change that to a YES. If `allowExcessShards=true` is set, the Decider will return YES regardless of filters: the test isn't proving anything.

DiannaHohensee replied:
> the new feedback suggests reverting to the deliberate setup

To make sure there's no confusion, I don't want you to revert the other randomization you added. The index shard number randomization is nice. This is my first review of your new node filter testing, and this particular line should not randomize the parameter supplied, for the reason given above.

zhubotang-wq commented:
> This is in good shape, pretty much done.

For scheduling and delivery purposes, we need to rely on code that is fully review-complete rather than ‘pretty good’ or ‘almost ready.’ To avoid confusion, could we hold off on those kinds of status descriptions until the changes are fully finalized?

DiannaHohensee commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,352 @@
+    private List<RoutingNode> indexTier;
+    private List<RoutingNode> searchIier;
+
+    private void setup(boolean hasDiscoveryNodeFilters, boolean allowExcessShards) {
```
Okay, I don't feel strongly about it. The method is very long, so it's a cheap way to make it easier for a reader to quickly understand the set up / have a preview to keep in mind.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Changes to support stateless test

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Changes to support stateless test

zhubotang-wq committed:
Changes to support stateless test

zhubotang-wq committed:
Changes to support stateless test

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Changes to support stateless test

nicktindall commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/routing/allocation/decider/WriteLoadConstraintDeciderIT.java:
```diff
@@ -604,6 +604,7 @@ private Settings enabledWriteLoadDeciderSettings(int utilizationThresholdPercent
                 WriteLoadConstraintSettings.WRITE_LOAD_DECIDER_QUEUE_LATENCY_THRESHOLD_SETTING.getKey(),
                 TimeValue.timeValueMillis(queueLatencyThresholdMillis)
             )
+            .put("cluster.routing.allocation.index_balance_decider.excess_shards", 1000)
```
We should be able to remove this now that the decider is disabled by default right?

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (indexBalanceConstraintSettings.isDeciderEnabled() == false || isStateless == false || hasNoFilters() == false) {
+            return Decision.single(Decision.Type.YES, NAME, "Decider is disabled.");
+        }
```
If it's disabled for stateful, I wonder if we could just configure it to be added in the stateless plugin? Then we wouldn't need to check `isStateless` every time.

nicktindall replied:
Also `hasNoFilters() == false` is harder to read than `filtersAreConfigured()` or similar. It's Friday here, I can't process a double-negative.

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+
+        if (node.node().getRoles().contains(SEARCH_ROLE) && shardRouting.primary()) {
+            return Decision.single(Decision.Type.YES, NAME, "A search node cannot own primary shards. Decider inactive.");
+        }
```
Are we duplicating logic here? this should be covered by `StatelessAllocationDecider`. Perhaps this is a substantial optimisation, in which case I guess it's OK, but if not I'd prefer to keep the deciders focused. 

If it is just here as an optimisation perhaps we could combine into a single check like `co.elastic.elasticsearch.stateless.allocation.StatelessAllocationDecider#canAllocateShardToNode` so it doesn't distract from the focus of this decider?

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+        return (clusterExcludeFilters == null || clusterExcludeFilters.hasNoFilters())
+            && (clusterIncludeFilters == null || clusterIncludeFilters.hasNoFilters())
+            && (clusterRequireFilters == null || clusterRequireFilters.hasNoFilters());
+    }
```
Do we need to keep the actual filters, or rather just track whether any are present here?

i.e could these be tracked as booleans instead e.g.

```java
clusterIncludeFilters = filters.size() > 0;
```

Doesn't seem like a big deal, but best to keep the state/logic minimal where we can, also might mean we can avoid adding `hasNoFilters` ?

nicktindall commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,352 @@
+        Settings.Builder builder = Settings.builder()
+            .put("stateless.enabled", "true")
+            .put("cluster.routing.allocation.index_balance_decider.enabled", "true")
+            .put("cluster.routing.allocation.index_balance_decider.excess_shards", excessShards);
```
Nit: I think better to refer to the `{settings constant}.getKey()` here

nicktindall commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    public void testCanAllocateUnderThresholdWithExcessShards() {
+        setup(false, true);
```
Nit: I think this is hard to read, especially outside of an IDE (methods with consecutive boolean params like this). Can we just pass in the settings or similar so it's clearer the starting point just by looking at the test?, or even replace the booleans with enums or something.

Could make a settings builder or something to reduce repetition if you think its worthwhile.

zhubotang-wq replied:
Agree. Fixed to use settings and clearly named parameters.

nicktindall reviewed (commented):
Looking good, a bunch of minor questions/comments from me

zhubotang-wq commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,167 @@
+                currentAllocation
+            );
+
+            logger.trace(explanation);
```
For this logging statement, 4 different reviewers offered different opinions. At this stage, I am inclined to leave it unchanged.

zhubotang-wq commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -497,6 +498,7 @@ public static Collection<AllocationDecider> createAllocationDeciders(
         addAllocationDecider(deciders, new ThrottlingAllocationDecider(clusterSettings));
         addAllocationDecider(deciders, new ShardsLimitAllocationDecider(clusterSettings));
         addAllocationDecider(deciders, new AwarenessAllocationDecider(settings, clusterSettings));
+        addAllocationDecider(deciders, new IndexBalanceAllocationDecider(settings, clusterSettings));
```
Ack. My previous replies agree this is a sound advice. At this stage, I am inclined to leave this refactoring to follow up iterations.

zhubotang-wq committed:
address feedbacks.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

nicktindall commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java:
```diff
@@ -0,0 +1,365 @@
+    private List<RoutingNode> indexTier;
+    private List<RoutingNode> searchIier;
+
+    private void setup(Settings settings) {
```
This isn't quite what I meant... what I was thinking is you could write utility methods for populating those settings e.g.


```java
    public Settings addRandomFilterSetting(Settings settings) {
           String setting = randomFrom(
                CLUSTER_ROUTING_REQUIRE_GROUP_PREFIX,
                CLUSTER_ROUTING_INCLUDE_GROUP_PREFIX,
                CLUSTER_ROUTING_EXCLUDE_GROUP_PREFIX
            );
            String attribute = randomFrom("_value", "name");
            String name = randomFrom("indexNodeOne", "indexNodeTwo", "searchNodeOne", "searchNodeTwo");
            String ip = randomFrom("192.168.0.1", "192.168.0.2", "192.168.7.1", "10.17.0.1");
            Settings.builder().put(settings)
                .put(setting + "." + attribute, attribute.equals("name") ? name : ip)
                .build();
    }

    public Settings allowExcessShards(Settings settings) {
        int excessShards = randomIntBetween(1, 5);

        return Settings.builder()
            .put(settings)           
            .put(IndexBalanceConstraintSettings.INDEX_BALANCE_DECIDER_EXCESS_SHARDS.getKey(), excessShards);
            .build();
    }
```


Then you could do e.g.

```java
     Settings settings = addRandomFilterSetting(Setting.EMPTY);
     settings = allowExcessShards(settings);
     setup(settings);

     //...
```

Then you just use the provided settings as a starting point for in the setup:

```java
    Settings Settings.builder().put(settings)
            .put("stateless.enabled", "true")
            .put(IndexBalanceConstraintSettings.INDEX_BALANCE_DECIDER_ENABLED_SETTING.getKey(), "true")
        .build();
```

That way the configuration of excess shards/random filters is just done in the places you need them and in the test rather than the setup method?

`Settings.builder().put(settings)` makes a builder with all the settings in the passed settings object copied.

nicktindall reviewed (commented):
I'm happy to approve once we're using `allocation.decision`, that's my only real comment that's not just opinion

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
address feedbacks.

zhubotang-wq commented:
> I'm happy to approve once we're using `allocation.decision`, that's my only real comment that's not just opinion

Made the following fixes. 

- Decision.Single has been replaced.
- add defensive if block for eligibleNodes.isEmpty() , retained the assertion based on previous discussion.
- add extra comments on 

>         final int threshold = (totalShards + eligibleNodes.size() - 1 + indexBalanceConstraintSettings.getExcessShards()) / 
>  eligibleNodes
>              .size();

- Integer division simpler to reason with than double.
- Add eligibleNodes.size() - 1 so that rounding down does not happen.
- eligibleNodes.size() - 1  is not so much that no automatic extra 1 shard allowed without touching ExcessShard setting

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDecider.java:
```diff
@@ -0,0 +1,172 @@
+        // The built-in "eligibleNodes.size() - 1" offers just enough buffer so threshold is not rounded down by integer division.
+        // But not too much so that threshold does not get an automatic 1 shard extra allowance.
+        final int threshold = (totalShards + eligibleNodes.size() - 1 + indexBalanceConstraintSettings.getExcessShards()) / eligibleNodes
+            .size();
```
Can we put this in a function called `ceiling_division` or something?

I see that it's a common method for doing integer ceiling division. There's also `Math.ceilDiv(int, int)` which I think would be even better. Leaving it as-is is just a bit mind boggling unless you're familiar with the trick.

zhubotang-wq replied:
The current form of calculation was suggested in an earlier feedback.

https://github.com/elastic/elasticsearch/pull/135875#discussion_r2485686023

The reviewer's rationale is as follows :

> Staying in int simplifies I think, though for common numbers, doubles do have exact precision, this helps me not speculate ;-). And adding the skew tolerance before division ensures that with tolerance 1 we get:
> 
> 2 shards, 2 nodes, allow 2 on each
> 3 shards, 2 nodes, allow 2 on each (prior version 3, but there is already 1 shard wiggle room).
> etc.

The threshold computation has gone through several iterations 


This calculation has evolved in several iterations. 

> final int threshold = (int) Math.ceil(idealAllocation * indexBalanceConstraintSettings.getLoadSkewTolerance());
> 
> final int threshold = (int) Math.ceil(idealAllocation) + indexBalanceConstraintSettings.getLoadSkewTolerance();
> 
> final int threshold = (totalShards + eligibleNodes.size() - 1 + indexBalanceConstraintSettings.getLoadSkewTolerance()) / eligibleNodes.size();

There’s a broad spectrum of perspectives on this topic. I’ve made an effort to incorporate as much feedback as possible, though aligning all perspectives has proven challenging. It’s been a bit difficult to understand the group’s decision-making dynamics in this process."

At this stage, I am inclined to keep it in its current form for this iteration.

nicktindall replied:
I don't think using the integer `ceilDiv` function from `java.lang.Math` contradicts any of the prior advice. For me, we want to have a pretty good reason to re-write logic that exists in the standard library and I don't think I see that here. It's just a straightforward readability/maintainability thing.

I think I do feel strongly about this one.

```java
final int threshold = Math.ceilDiv(totalShards + indexBalanceConstraintSettings.getExcessShards(), eligibleNodes.size());
```

is much easier to understand than

```java
final int threshold = (totalShards + eligibleNodes.size() - 1 + indexBalanceConstraintSettings.getExcessShards()) / eligibleNodes
            .size();
```

`Math.ceilDiv(int n, int d)` seems to be equivalent to `n + d - 1 / d` except that the former is very nicely documented. I don't think @henningandersen would disagree?

henningandersen replied:
yeah, I agree to this, using a standard function is better.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
address feedbacks.

zhubotang-wq committed:
Merge remote-tracking branch 'origin/12080-index-shard-count-allocation-decider' into 12080-index-shard-count-allocation-decider

zhubotang-wq committed:
address feedbacks.

zhubotang-wq committed:
address feedbacks.

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider

nicktindall reviewed (approved):
This LGTM

zhubotang-wq committed:
Merge branch 'main' into 12080-index-shard-count-allocation-decider
